### PR TITLE
`<bit>`: `bit_ceil`(T(-1)) is not a constant expression

### LINKED ISSUE #1595: <bit>: bit_ceil(T(-1)) should not be a constant expression
Per [[bit.pow.two]/8](http://eel.is/c++draft/bit.pow.two#8) `std::bit_ceil(meow)` is not a core constant expression when the smallest power of two greater than or equal to `meow` is not representable by the type `decltype(meow)`. This is not the case for all supported types and all supported compilers with our implementation as witnessed by this ill-formed TU:
```c++
#include <bit>

using UC = unsigned char;
using US = unsigned short;
using UI = unsigned int;
using UL = unsigned long;
using ULL = unsigned long long;

// All static_asserts are ill-formed since their
// arguments are not constant expressions.

// None of Clang/EDG/MSVC diagnose these:
static_assert(((void)std::bit_ceil(UC(-1)), true));
static_assert(((void)std::bit_ceil(US(-1)), true));

// Clang and EDG, but not MSVC, diagnose these:
static_assert(((void)std::bit_ceil(UI(-1)), true));
static_assert(((void)std::bit_ceil(UL(-1)), true));
static_assert(((void)std::bit_ceil(ULL(-1)), true));
```

There's almost certainly an MSVC compiler bug here, and probably a library bug as well.

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #1595 

https://gcc.godbolt.org/z/68MGqPG8z

`bit_ceil.fail.cpp` is already unskipped.

https://github.com/microsoft/STL/pull/1920/files#diff-cc6f92924b9b137833cf021a8871d88049f0dd2268a6c2749a734c81503f6cf7L435-L437

### CODE REVIEW

**Comment by fsb4000** [comment]
based on MSVC implementation + https://en.cppreference.com/w/cpp/numeric/bit_ceil

File: stl/inc/bit [hunk 1]
```diff
@@ -44,11 +44,17 @@ _NODISCARD constexpr bool has_single_bit(const _Ty _Val) noexcept {
...
+        return _Ty{1};
     }
 
-    return static_cast<_Ty>(_Ty{1} << (numeric_limits<_Ty>::digits - _STD countl_zero(static_cast<_Ty>(_Val - 1))));
+    const int _Num = numeric_limits<_Ty>::digits - _STD countl_zero(static_cast<_Ty>(_Val - 1));
+    if constexpr (sizeof(_Ty) >= sizeof(unsigned int)) {
+        return _Ty{1} << _Num;
+    } else { // for types subject to integral promotion
+        constexpr int _Offset_for_ub = numeric_limits<unsigned int>::digits - numeric_limits<_Ty>::digits;
+        return _Ty(1u << (_Num + _Offset_for_ub) >> _Offset_for_ub);
```
Comment (AdamBucior) [inline stl/inc/bit:56]: We should do it only if `is_constant_evaluated`.

File: stl/inc/bit [hunk 2]
```diff
@@ -44,11 +44,21 @@ _NODISCARD constexpr bool has_single_bit(const _Ty _Val) noexcept {
...
+        if constexpr (sizeof(_Ty) >= sizeof(unsigned int)) {
+            return _Ty{1} << _Num;
+        } else { // for types subject to integral promotion
+            constexpr int _Offset_for_ub = numeric_limits<unsigned int>::digits - numeric_limits<_Ty>::digits;
+            return static_cast<_Ty>(1u << (_Num + _Offset_for_ub) >> _Offset_for_ub);
+        }
     }
 
-    return static_cast<_Ty>(_Ty{1} << (numeric_limits<_Ty>::digits - _STD countl_zero(static_cast<_Ty>(_Val - 1))));
+    return static_cast<_Ty>(_Ty{1} << _Num);
```
Comment (miscco) [inline stl/inc/bit:61]: I would prefer if we could do the `is_constant_evaluated` branch inside the `if constexpr`

```cpp
    if constexpr (sizeof(_Ty) >= sizeof(unsigned int)) {
        return _Ty{1} << _Num;
    } else { // for types subject to integral promotion
        if (_STD is_constant_evaluated()) {
            constexpr int _Offset_for_ub = numeric_limits<unsigned int>::digits - numeric_limits<_Ty>::digits;
            return static_cast<_Ty>(1u << (_Num + _Offset_for_ub) >> _Offset_for_ub);
       } else {
           return _Ty{1} << _Num;
       }
    }
```

That way we do not need to bother the compiler with the whole `is_constant_evaluated` when we do not promote

**Comment by barcharcraz** [comment]
We should probably reduce the compiler bug here so we can simplify things again when it's fixed.

**Comment by fsb4000** [comment]
@barcharcraz This is a library bug. Not a compiler one. See: https://gcc.godbolt.org/z/o6Wx395zM

File: stl/inc/bit [hunk 3]
```diff
@@ -44,11 +44,21 @@ _NODISCARD constexpr bool has_single_bit(const _Ty _Val) noexcept {
...
     }
 
-    return static_cast<_Ty>(_Ty{1} << (numeric_limits<_Ty>::digits - _STD countl_zero(static_cast<_Ty>(_Val - 1))));
+    const int _Num = numeric_limits<_Ty>::digits - _STD countl_zero(static_cast<_Ty>(_Val - 1));
+    if constexpr (sizeof(_Ty) >= sizeof(unsigned int)) {
+        return _Ty{1} << _Num;
+    } else { // for types subject to integral promotion
+        if (_STD is_constant_evaluated()) {
+            constexpr int _Offset_for_ub = numeric_limits<unsigned int>::digits - numeric_limits<_Ty>::digits;
+            return static_cast<_Ty>(1u << (_Num + _Offset_for_ub) >> _Offset_for_ub);
```
Comment (barcharcraz) [inline stl/inc/bit:57]: An explanation of exactly why we're doing this here would be good for the future, so we don't regress.

**Comment by barcharcraz** [comment]
Why are we not failing bit_ceil..fail.cpp after the toolset update?

**Review by barcharcraz — APPROVED** [review]

**Comment by fsb4000** [comment]
@barcharcraz Because we raise error for the 3 `static_assert`s : https://github.com/llvm/llvm-project/blob/e23dce6c974477d3476bee5256a83a9aaea8eae4/libcxx/test/std/numerics/bit/bit.pow.two/bit_ceil.fail.cpp#L39-L41

But we don't raise error for the 2`static_assert`'s: https://github.com/llvm/llvm-project/blob/e23dce6c974477d3476bee5256a83a9aaea8eae4/libcxx/test/std/numerics/bit/bit.pow.two/bit_ceil.fail.cpp#L37-L38

And our testing system only checks that the compilation was unsuccessful (without checking the reasons)

**Comment by barcharcraz** [comment]
> @barcharcraz Because we raise error for the 3 `static_assert`s : https://github.com/llvm/llvm-project/blob/e23dce6c974477d3476bee5256a83a9aaea8eae4/libcxx/test/std/numerics/bit/bit.pow.two/bit_ceil.fail.cpp#L39-L41
> 
> But we don't raise error for the 2`static_assert`'s: https://github.com/llvm/llvm-project/blob/e23dce6c974477d3476bee5256a83a9aaea8eae4/libcxx/test/std/numerics/bit/bit.pow.two/bit_ceil.fail.cpp#L37-L38
> 
> And our testing system only checks that the compilation was unsuccessful (without checking the reasons)

it would be good to correct the test to actually check for all the things, and be a normal test instead of an expected fail (we could also add such a test to our test suite)

That doesn't need to be in this PR though.

**Review by barcharcraz — APPROVED** [review]

**Review by StephanTLavavej — COMMENTED** [review]
Thanks for fixing this! Although this is all correct, I believe that it could be refactored to be clearer - since I have several overlapping changes in mind, I'll go ahead and push proposed commits.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
@fsb4000 @barcharcraz I went ahead and pushed a refactoring - I hope that's okay. Here's what the diagnostic for a constexpr precondition violation looks like now:

```
C:\GitHub\STL\out\build\x64>type meow.cpp
```
```cpp
#include <bit>

static_assert(std::bit_ceil(static_cast<unsigned short>(0xFFFF)) > 0u);

int main() {}
```
```
C:\GitHub\STL\out\build\x64>cl /EHsc /nologo /W4 /std:c++latest /c meow.cpp
meow.cpp
meow.cpp(3): error C2131: expression did not evaluate to a constant
C:\GitHub\STL\out\build\x64\out\inc\bit(67): note: failure was caused by call of undefined function or one not declared 'constexpr'
C:\GitHub\STL\out\build\x64\out\inc\bit(67): note: see usage of 'std::_Precondition_violation_in_bit_ceil'

C:\GitHub\STL\out\build\x64>clang-cl /EHsc /nologo /W4 /std:c++latest /c meow.cpp
meow.cpp(3,15): error: static_assert expression is not an integral constant expression
static_assert(std::bit_ceil(static_cast<unsigned short>(0xFFFF)) > 0u);
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\GitHub\STL\out\build\x64\out\inc\bit(67,17): note: non-constexpr function '_Precondition_violation_in_bit_ceil'
      cannot be used in a constant expression
                _Precondition_violation_in_bit_ceil();
                ^
meow.cpp(3,15): note: in call to 'bit_ceil(65535)'
static_assert(std::bit_ceil(static_cast<unsigned short>(0xFFFF)) > 0u);
              ^
C:\GitHub\STL\out\build\x64\out\inc\bit(45,13): note: declared here
inline void _Precondition_violation_in_bit_ceil() noexcept {}
            ^
1 error generated.
```

I believe that this will be easier for users to understand than the constexpr evaluator rejecting the "shift with an offset" logic.

**Comment by StephanTLavavej** [comment]
Thanks again for fixing this bug! :bug: :heavy_check_mark: :smile_cat:

**Pull request merged**